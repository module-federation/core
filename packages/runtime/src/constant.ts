export const DEFAULT_SCOPE = 'default';
export const DEFAULT_REMOTE_TYPE = 'global';
