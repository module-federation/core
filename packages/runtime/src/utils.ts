import { FederationHost } from '@module-federation/runtime-core';
import { CurrentGlobal } from '@module-federation/runtime-core';

// injected by bundler, so it can not use runtime-core stuff
export function getBuilderId(): string {
  //@ts-ignore
  return typeof FEDERATION_BUILD_IDENTIFIER !== 'undefined'
    ? //@ts-ignore
      FEDERATION_BUILD_IDENTIFIER
    : '';
}

export function getGlobalFederationInstance(
  name: string,
  version: string | undefined,
): FederationHost | undefined {
  const buildId = getBuilderId();
  return CurrentGlobal.__FEDERATION__.__INSTANCES__.find((GMInstance) => {
    if (buildId && GMInstance.options.id === getBuilderId()) {
      return true;
    }

    if (
      GMInstance.options.name === name &&
      !GMInstance.options.version &&
      !version
    ) {
      return true;
    }

    if (
      GMInstance.options.name === name &&
      version &&
      GMInstance.options.version === version
    ) {
      return true;
    }
    return false;
  });
}
