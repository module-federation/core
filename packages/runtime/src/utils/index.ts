export * from './env';
export * from './tool';
export * from './manifest';
export * from './logger';
export * from './plugin';
