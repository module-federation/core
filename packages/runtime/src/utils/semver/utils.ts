// fork from https://github.com/originjs/vite-plugin-federation/blob/v1.1.12/packages/lib/src/utils/semver/index.ts
// Copyright (c)
// vite-plugin-federation is licensed under Mulan PSL v2.
// You can use this software according to the terms and conditions of the Mulan PSL v2.
// You may obtain a copy of Mulan PSL v2 at:
//      http://license.coscl.org.cn/MulanPSL2
// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
// See the Mulan PSL v2 for more details.

import { comparator } from './constants';

export function parseRegex(source: string): RegExp {
  return new RegExp(source);
}

export function isXVersion(version: string): boolean {
  return !version || version.toLowerCase() === 'x' || version === '*';
}

export function pipe<TArgs extends any[], R1, R2, R3, R4, R5, R6, R7>(
  f1: (...args: TArgs) => R1,
  f2: (a: R1) => R2,
  f3: (a: R2) => R3,
  f4: (a: R3) => R4,
  f5: (a: R4) => R5,
  f6: (a: R5) => R6,
  f7: (a: R6) => R7,
): (...args: TArgs) => R7;
export function pipe<TArgs extends any[], R1, R2, R3, R4, R5, R6>(
  f1: (...args: TArgs) => R1,
  f2: (a: R1) => R2,
  f3: (a: R2) => R3,
  f4: (a: R3) => R4,
  f5: (a: R4) => R5,
  f6: (a: R5) => R6,
): (...args: TArgs) => R6;
export function pipe<TArgs extends any[], R1, R2, R3, R4, R5>(
  f1: (...args: TArgs) => R1,
  f2: (a: R1) => R2,
  f3: (a: R2) => R3,
  f4: (a: R3) => R4,
  f5: (a: R4) => R5,
): (...args: TArgs) => R5;
export function pipe<TArgs extends any[], R1, R2, R3, R4>(
  f1: (...args: TArgs) => R1,
  f2: (a: R1) => R2,
  f3: (a: R2) => R3,
  f4: (a: R3) => R4,
): (...args: TArgs) => R4;
export function pipe<TArgs extends any[], R1, R2, R3>(
  f1: (...args: TArgs) => R1,
  f2: (a: R1) => R2,
  f3: (a: R2) => R3,
): (...args: TArgs) => R3;
export function pipe<TArgs extends any[], R1, R2>(
  f1: (...args: TArgs) => R1,
  f2: (a: R1) => R2,
): (...args: TArgs) => R2;
export function pipe<TArgs extends any[], R1>(
  f1: (...args: TArgs) => R1,
): (...args: TArgs) => R1;
export function pipe(...fns: ((params: any) => any)[]) {
  return (x: unknown): any => fns.reduce((v, f) => f(v), x);
}

export function extractComparator(
  comparatorString: string,
): RegExpMatchArray | null {
  return comparatorString.match(parseRegex(comparator));
}

export function combineVersion(
  major: string,
  minor: string,
  patch: string,
  preRelease: string,
): string {
  const mainVersion = `${major}.${minor}.${patch}`;

  if (preRelease) {
    return `${mainVersion}-${preRelease}`;
  }

  return mainVersion;
}
