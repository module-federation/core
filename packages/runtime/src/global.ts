import { FederationHost } from './core';
import {
  RemoteEntryExports,
  GlobalShareScopeMap,
  Remote,
  Optional,
} from './type';
import { getFMId } from './utils/tool';
import {
  GlobalModuleInfo,
  ModuleInfo,
  isDebugMode,
} from '@module-federation/sdk';
import { getBuilderId } from './utils/env';
import { warn } from './utils/logger';
import { FederationRuntimePlugin } from './type/plugin';

export interface Federation {
  __GLOBAL_PLUGIN__: Array<FederationRuntimePlugin>;
  __DEBUG_CONSTRUCTOR_VERSION__?: string;
  moduleInfo: GlobalModuleInfo;
  __DEBUG_CONSTRUCTOR__?: typeof FederationHost;
  __INSTANCES__: Array<FederationHost>;
  __SHARE__: GlobalShareScopeMap;
  __MANIFEST_LOADING__: Record<string, Promise<ModuleInfo>>;
  __PRELOADED_MAP__: Map<string, boolean>;
}

export const nativeGlobal: typeof global = (() => {
  try {
    return new Function('return this')();
  } catch {
    return globalThis;
  }
})() as typeof global;

export const Global = nativeGlobal;

declare global {
  // eslint-disable-next-line no-var
  var __FEDERATION__: Federation,
    __VMOK__: Federation,
    // eslint-disable-next-line no-var
    __GLOBAL_LOADING_REMOTE_ENTRY__: Record<
      string,
      undefined | Promise<RemoteEntryExports | void>
    >;
}

function definePropertyGlobalVal(
  target: typeof globalThis,
  key: string,
  val: any,
) {
  Object.defineProperty(target, key, {
    value: val,
    configurable: false,
    writable: true,
  });
}

function includeOwnProperty(target: typeof globalThis, key: string) {
  return Object.hasOwnProperty.call(target, key);
}

// This section is to prevent encapsulation by certain microfrontend frameworks. Due to reuse policies, sandbox escapes.
// The sandbox in the microfrontend does not replicate the value of 'configurable'.
// If there is no loading content on the global object, this section defines the loading object.
if (!includeOwnProperty(globalThis, '__GLOBAL_LOADING_REMOTE_ENTRY__')) {
  definePropertyGlobalVal(globalThis, '__GLOBAL_LOADING_REMOTE_ENTRY__', {});
}

export const globalLoading = globalThis.__GLOBAL_LOADING_REMOTE_ENTRY__;

function setGlobalDefaultVal(target: typeof globalThis) {
  if (
    includeOwnProperty(target, '__VMOK__') &&
    !includeOwnProperty(target, '__FEDERATION__')
  ) {
    definePropertyGlobalVal(target, '__FEDERATION__', target.__VMOK__);
  }

  if (!includeOwnProperty(target, '__FEDERATION__')) {
    definePropertyGlobalVal(target, '__FEDERATION__', {
      __GLOBAL_PLUGIN__: [],
      __INSTANCES__: [],
      moduleInfo: {},
      __SHARE__: {},
      __MANIFEST_LOADING__: {},
      __PRELOADED_MAP__: new Map(),
    });

    definePropertyGlobalVal(target, '__VMOK__', target.__FEDERATION__);
  }

  target.__FEDERATION__.__GLOBAL_PLUGIN__ ??= [];
  target.__FEDERATION__.__INSTANCES__ ??= [];
  target.__FEDERATION__.moduleInfo ??= {};
  target.__FEDERATION__.__SHARE__ ??= {};
  target.__FEDERATION__.__MANIFEST_LOADING__ ??= {};
  target.__FEDERATION__.__PRELOADED_MAP__ ??= new Map();
}

setGlobalDefaultVal(globalThis);
setGlobalDefaultVal(nativeGlobal);

export function resetFederationGlobalInfo(): void {
  globalThis.__FEDERATION__.__GLOBAL_PLUGIN__ = [];
  globalThis.__FEDERATION__.__INSTANCES__ = [];
  globalThis.__FEDERATION__.moduleInfo = {};
  globalThis.__FEDERATION__.__SHARE__ = {};
  globalThis.__FEDERATION__.__MANIFEST_LOADING__ = {};
}

export function getGlobalFederationInstance(
  name: string,
  version: string | undefined,
): FederationHost | undefined {
  const buildId = getBuilderId();
  return globalThis.__FEDERATION__.__INSTANCES__.find((GMInstance) => {
    if (buildId && GMInstance.options.id === getBuilderId()) {
      return true;
    }

    if (
      GMInstance.options.name === name &&
      !GMInstance.options.version &&
      !version
    ) {
      return true;
    }

    if (
      GMInstance.options.name === name &&
      version &&
      GMInstance.options.version === version
    ) {
      return true;
    }
    return false;
  });
}

export function setGlobalFederationInstance(
  FederationInstance: FederationHost,
): void {
  globalThis.__FEDERATION__.__INSTANCES__.push(FederationInstance);
}

export function getGlobalFederationConstructor():
  | typeof FederationHost
  | undefined {
  return globalThis.__FEDERATION__.__DEBUG_CONSTRUCTOR__;
}

export function setGlobalFederationConstructor(
  FederationConstructor: typeof FederationHost | undefined,
  isDebug = isDebugMode(),
): void {
  if (isDebug) {
    globalThis.__FEDERATION__.__DEBUG_CONSTRUCTOR__ = FederationConstructor;
    globalThis.__FEDERATION__.__DEBUG_CONSTRUCTOR_VERSION__ = __VERSION__;
  }
}

// eslint-disable-next-line @typescript-eslint/ban-types
export function getInfoWithoutType<T extends object>(
  target: T,
  key: keyof T,
): { value: T[keyof T] | undefined; key: string } {
  if (typeof key === 'string') {
    const keyRes = target[key];
    if (keyRes) {
      return {
        value: target[key],
        key: key as string,
      };
    } else {
      const targetKeys = Object.keys(target);
      for (const targetKey of targetKeys) {
        const [targetTypeOrName, _] = targetKey.split(':');
        const nKey = `${targetTypeOrName}:${key}` as unknown as keyof T;
        const typeWithKeyRes = target[nKey];
        if (typeWithKeyRes) {
          return {
            value: typeWithKeyRes,
            key: nKey as string,
          };
        }
      }
      return {
        value: undefined,
        key: key as string,
      };
    }
  } else {
    throw new Error('key must be string');
  }
}

export const getGlobalSnapshot = (): GlobalModuleInfo =>
  nativeGlobal.__FEDERATION__.moduleInfo;

export const getTargetSnapshotInfoByModuleInfo = (
  moduleInfo: Optional<Remote, 'alias'>,
  snapshot: GlobalModuleInfo,
): GlobalModuleInfo[string] | undefined => {
  // Check if the remote is included in the hostSnapshot
  const moduleKey = getFMId(moduleInfo);
  const getModuleInfo = getInfoWithoutType(snapshot, moduleKey).value;

  // The remoteSnapshot might not include a version
  if (
    getModuleInfo &&
    !getModuleInfo.version &&
    'version' in moduleInfo &&
    moduleInfo['version']
  ) {
    getModuleInfo.version = moduleInfo['version'];
  }

  if (getModuleInfo) {
    return getModuleInfo;
  }

  // If the remote is not included in the hostSnapshot, deploy a micro app snapshot
  if ('version' in moduleInfo && moduleInfo['version']) {
    const { version, ...resModuleInfo } = moduleInfo;
    const moduleKeyWithoutVersion = getFMId(resModuleInfo);
    const getModuleInfoWithoutVersion = getInfoWithoutType(
      nativeGlobal.__FEDERATION__.moduleInfo,
      moduleKeyWithoutVersion,
    ).value;

    if (getModuleInfoWithoutVersion?.version === version) {
      return getModuleInfoWithoutVersion;
    }
  }

  return;
};

export const getGlobalSnapshotInfoByModuleInfo = (
  moduleInfo: Optional<Remote, 'alias'>,
): GlobalModuleInfo[string] | undefined =>
  getTargetSnapshotInfoByModuleInfo(
    moduleInfo,
    nativeGlobal.__FEDERATION__.moduleInfo,
  );

export const setGlobalSnapshotInfoByModuleInfo = (
  remoteInfo: Remote,
  moduleDetailInfo: GlobalModuleInfo[string],
): GlobalModuleInfo => {
  const moduleKey = getFMId(remoteInfo);
  nativeGlobal.__FEDERATION__.moduleInfo[moduleKey] = moduleDetailInfo;
  return nativeGlobal.__FEDERATION__.moduleInfo;
};

export const addGlobalSnapshot = (
  moduleInfos: GlobalModuleInfo,
): (() => void) => {
  nativeGlobal.__FEDERATION__.moduleInfo = {
    ...nativeGlobal.__FEDERATION__.moduleInfo,
    ...moduleInfos,
  };
  return () => {
    const keys = Object.keys(moduleInfos);
    for (const key of keys) {
      delete nativeGlobal.__FEDERATION__.moduleInfo[key];
    }
  };
};

export const getRemoteEntryExports = (
  name: string,
  globalName: string | undefined,
): {
  remoteEntryKey: string;
  entryExports: RemoteEntryExports | undefined;
} => {
  const remoteEntryKey = globalName || `__FEDERATION_${name}:custom__`;
  const entryExports = (globalThis as any)[remoteEntryKey];
  return {
    remoteEntryKey,
    entryExports,
  };
};

// This function is used to register global plugins.
// It iterates over the provided plugins and checks if they are already registered.
// If a plugin is not registered, it is added to the global plugins.
// If a plugin is already registered, a warning message is logged.
export const registerGlobalPlugins = (
  plugins: Array<FederationRuntimePlugin>,
): void => {
  const { __GLOBAL_PLUGIN__ } = nativeGlobal.__FEDERATION__;

  plugins.forEach((plugin) => {
    if (__GLOBAL_PLUGIN__.findIndex((p) => p.name === plugin.name) === -1) {
      __GLOBAL_PLUGIN__.push(plugin);
    } else {
      warn(`The plugin ${plugin.name} has been registered.`);
    }
  });
};

export const getGlobalHostPlugins = (): Array<FederationRuntimePlugin> =>
  nativeGlobal.__FEDERATION__.__GLOBAL_PLUGIN__;

export const getPreloaded = (id: string) =>
  globalThis.__FEDERATION__.__PRELOADED_MAP__.get(id);

export const setPreloaded = (id: string) =>
  globalThis.__FEDERATION__.__PRELOADED_MAP__.set(id, true);
