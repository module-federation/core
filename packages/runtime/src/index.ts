import { FederationHost } from './core';
import {
  getGlobalFederationInstance,
  getGlobalFederationConstructor,
  setGlobalFederationInstance,
  setGlobalFederationConstructor,
} from './global';
import { UserOptions, FederationRuntimePlugin } from './type';
import { assert } from './utils/logger';

export { FederationHost } from './core';
export { registerGlobalPlugins } from './global';
export { getRemoteEntry, getRemoteInfo } from './utils';
export { loadScript, loadScriptNode } from '@module-federation/sdk';

export type { Federation } from './global';
export type { FederationRuntimePlugin };

let FederationInstance: FederationHost | null = null;
export function init(options: UserOptions): FederationHost {
  // Retrieve the same instance with the same name
  const instance = getGlobalFederationInstance(options.name, options.version);
  if (!instance) {
    // Retrieve debug constructor
    const FederationConstructor =
      getGlobalFederationConstructor() || FederationHost;
    FederationInstance = new FederationConstructor(options);
    setGlobalFederationInstance(FederationInstance);
    return FederationInstance;
  } else {
    // Merge options
    instance.initOptions(options);
    if (!FederationInstance) {
      FederationInstance = instance;
    }
    return instance;
  }
}

export function loadRemote<T>(
  ...args: Parameters<FederationHost['loadRemote']>
): Promise<T | null> {
  assert(FederationInstance, 'Please call init first');
  const loadRemote: typeof FederationInstance.loadRemote<T> =
    FederationInstance.loadRemote;
  // eslint-disable-next-line prefer-spread
  return loadRemote.apply(FederationInstance, args);
}

export function loadShare<T>(
  ...args: Parameters<FederationHost['loadShare']>
): Promise<false | (() => T | undefined)> {
  assert(FederationInstance, 'Please call init first');
  // eslint-disable-next-line prefer-spread
  const loadShare: typeof FederationInstance.loadShare<T> =
    FederationInstance.loadShare;
  return loadShare.apply(FederationInstance, args);
}

export function loadShareSync<T>(
  ...args: Parameters<FederationHost['loadShareSync']>
): () => T | never {
  assert(FederationInstance, 'Please call init first');
  const loadShareSync: typeof FederationInstance.loadShareSync<T> =
    FederationInstance.loadShareSync;
  // eslint-disable-next-line prefer-spread
  return loadShareSync.apply(FederationInstance, args);
}

export function preloadRemote(
  ...args: Parameters<FederationHost['preloadRemote']>
): ReturnType<FederationHost['preloadRemote']> {
  assert(FederationInstance, 'Please call init first');
  // eslint-disable-next-line prefer-spread
  return FederationInstance.preloadRemote.apply(FederationInstance, args);
}

export function registerRemotes(
  ...args: Parameters<FederationHost['registerRemotes']>
): ReturnType<FederationHost['registerRemotes']> {
  assert(FederationInstance, 'Please call init first');
  // eslint-disable-next-line prefer-spread
  return FederationInstance.registerRemotes.apply(FederationInstance, args);
}

export function registerPlugins(
  ...args: Parameters<FederationHost['registerPlugins']>
): ReturnType<FederationHost['registerRemotes']> {
  assert(FederationInstance, 'Please call init first');
  // eslint-disable-next-line prefer-spread
  return FederationInstance.registerPlugins.apply(FederationInstance, args);
}

// Inject for debug
setGlobalFederationConstructor(FederationHost);
