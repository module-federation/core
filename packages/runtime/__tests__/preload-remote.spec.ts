import { describe, it } from 'vitest';
import { init } from '../src/index';
import { mockStaticServer } from './mock/utils';
import { Global, addGlobalSnapshot } from '../src/global';
interface LinkInfo {
  type: string;
  href: string;
  rel: string;
}

interface ScriptInfo {
  src: string;
  crossorigin: string;
}

function getLinkInfos(): Array<LinkInfo> {
  const links = document.querySelectorAll('link');
  return Array.from(links).map((link) => ({
    type: link.getAttribute('as') || '',
    href: link.getAttribute('href') || '',
    rel: link.getAttribute('rel') || '',
  }));
}
function getScriptInfos(): Array<ScriptInfo> {
  const scripts = document.querySelectorAll('script');
  return Array.from(scripts).map((script) => ({
    src: script.getAttribute('src') || '',
    crossorigin: script.getAttribute('crossorigin') || '',
  }));
}
function getPreloadElInfos() {
  return {
    links: getLinkInfos(),
    scripts: getScriptInfos(),
  };
}
describe('preload-remote inBrowser', () => {
  mockStaticServer({
    baseDir: __dirname,
    filterKeywords: [],
    basename: 'http://localhost:1111/',
  });
  const mockSnapshot = {
    '@federation/preload-remote': {
      buildVersion: 'custom',
      publicPath: 'http://localhost:1111/resources/preload/preload-resource/',
      remoteEntry: 'federation-remote-entry.js',
      remotesInfo: {
        '@federation/sub1': { matchedVersion: '1.0.2' },
        '@federation/sub2': { matchedVersion: '1.0.3' },
        '@federation/sub3': { matchedVersion: '1.0.3' },
      },
    },
    '@federation/sub1:1.0.2': {
      buildVersion: 'custom',
      modules: [
        {
          moduleName: 'button',
          assets: {
            css: {
              sync: ['button.sync.css', 'ignore.sync.css'],
              async: ['button.async.css'],
            },
            js: {
              sync: ['button.sync.js'],
              async: ['button.async.js'],
            },
          },
        },
      ],
      publicPath: 'http://localhost:1111/resources/preload/preload-resource/',
      remoteEntry: 'federation-remote-entry.js',
      remotesInfo: {
        '@federation/sub1-button': { matchedVersion: '1.0.3' },
        '@federation/sub1-add': { matchedVersion: '1.0.3' },
      },
    },
    '@federation/sub1-button:1.0.3': {
      buildVersion: 'custom',
      modules: [
        {
          moduleName: 'button',
          assets: {
            css: { sync: [], async: [] },
            js: {
              sync: ['sub1-button/button.sync.js'],
              async: ['sub1-button/button.async.js'],
            },
          },
        },
      ],
      publicPath: 'http://localhost:1111/resources/preload/preload-resource/',
      remoteEntry: 'sub1-button/federation-remote-entry.js',
    },
    '@federation/sub1-add:1.0.3': {
      buildVersion: 'custom',
      modules: [
        {
          moduleName: 'add',
          assets: {
            css: { sync: [], async: [] },
            js: {
              sync: ['sub1-add/add.sync.js'],
              async: ['sub1-add/add.async.js'],
            },
          },
        },
      ],
      publicPath: 'http://localhost:1111/resources/preload/preload-resource/',
      remoteEntry: 'sub1-button/federation-remote-entry.js',
    },
    '@federation/sub2:1.0.3': {
      buildVersion: 'custom',
      modules: [
        {
          moduleName: 'button',
          assets: {
            css: {
              sync: ['sub2/button.sync.css'],
              async: ['sub2/button.async.css'],
            },
            js: {
              sync: ['sub2/button.sync.js'],
              async: ['sub2/button.async.js'],
            },
          },
        },
      ],
      publicPath: 'http://localhost:1111/resources/preload/preload-resource/',
      remoteEntry: 'sub2/federation-remote-entry.js',
      remotesInfo: {
        '@federation/sub2-button': { matchedVersion: '1.0.3' },
        '@federation/sub2-add': { matchedVersion: '1.0.3' },
      },
    },
    '@federation/sub2-button:1.0.3': {
      buildVersion: 'custom',
      modules: [
        {
          moduleName: 'button',
          assets: {
            css: { sync: [], async: [] },
            js: {
              sync: ['sub2-button/button.sync.js'],
              async: ['sub2-button/button.async.js'],
            },
          },
        },
      ],
      publicPath: 'http://localhost:1111/resources/preload/preload-resource/',
      remoteEntry: 'sub2-button/federation-remote-entry.js',
    },
    '@federation/sub2-add:1.0.3': {
      buildVersion: 'custom',
      modules: [
        {
          moduleName: 'add',
          assets: {
            css: { sync: [], async: [] },
            js: {
              sync: ['sub2-add/add.sync.js'],
              async: ['sub2-add/add.async.js'],
            },
          },
        },
      ],
      publicPath: 'http://localhost:1111/resources/preload/preload-resource/',
      remoteEntry: 'sub2-button/federation-remote-entry.js',
    },
    '@federation/sub3:1.0.3': {
      buildVersion: 'custom',
      modules: [
        {
          moduleName: 'button',
          assets: {
            css: { sync: [], async: [] },
            js: { sync: ['sub3/button.sync.js'], async: [] },
          },
        },
        {
          moduleName: 'add',
          assets: {
            css: { sync: [], async: [] },
            js: { sync: ['sub3/add.sync.js'], async: [] },
          },
        },
      ],
      publicPath: 'http://localhost:1111/resources/preload/preload-resource/',
      remoteEntry: 'sub3/federation-remote-entry.js',
    },
  };
  beforeEach(() => {
    document.head.innerHTML = '';
    document.body.innerHTML = '';
    Global.__FEDERATION__.__PRELOADED_MAP__.clear();
  });
  const FMInstance = init({
    name: '@federation/preload-remote',
    remotes: [
      { name: '@federation/sub1', version: '1.0.2' },
      { name: '@federation/sub2', version: '1.0.3' },
      { name: '@federation/sub3', version: '1.0.3' },
    ],
    plugins: [
      {
        name: 'preload-resource-inbrowser',
        beforeInit(args) {
          args.options.inBrowser = true;
          return args;
        },
      },
    ],
  });
  it('1 preload with default config', async () => {
    const reset = addGlobalSnapshot(mockSnapshot);
    expect(Global.__FEDERATION__.__PRELOADED_MAP__.size).toBe(0);
    await FMInstance.preloadRemote([
      {
        nameOrAlias: '@federation/sub1',
        filter(assetUrl) {
          return assetUrl.indexOf('ignore') === -1;
        },
        depsRemote: [{ nameOrAlias: '@federation/sub1-button' }],
      },
    ]);

    expect(getPreloadElInfos()).toMatchSnapshot();
    expect(Global.__FEDERATION__.__PRELOADED_MAP__.size).toBe(2);
    expect(
      Global.__FEDERATION__.__PRELOADED_MAP__.get('@federation/sub1/button'),
    ).toBe(true);
    expect(
      Global.__FEDERATION__.__PRELOADED_MAP__.get(
        '@federation/sub1-button/button',
      ),
    ).toBe(true);
    reset();
  });

  it('2 preload with all config ', async () => {
    const reset = addGlobalSnapshot(mockSnapshot);
    expect(Global.__FEDERATION__.__PRELOADED_MAP__.size).toBe(0);
    await FMInstance.preloadRemote([
      {
        nameOrAlias: '@federation/sub2',
        resourceCategory: 'all',
      },
    ]);

    expect(getPreloadElInfos()).toMatchSnapshot();
    expect(Global.__FEDERATION__.__PRELOADED_MAP__.size).toBe(3);
    expect(
      Global.__FEDERATION__.__PRELOADED_MAP__.get('@federation/sub2/button'),
    ).toBe(true);
    expect(
      Global.__FEDERATION__.__PRELOADED_MAP__.get(
        '@federation/sub2-button/button',
      ),
    ).toBe(true);
    expect(
      Global.__FEDERATION__.__PRELOADED_MAP__.get('@federation/sub2-add/add'),
    ).toBe(true);
    reset();
  });

  it('3 preload with expose config ', async () => {
    const reset = addGlobalSnapshot(mockSnapshot);

    expect(Global.__FEDERATION__.__PRELOADED_MAP__.size).toBe(0);
    await FMInstance.preloadRemote([
      {
        nameOrAlias: '@federation/sub3',
        resourceCategory: 'all',
        exposes: ['add'],
      },
    ]);
    expect(getPreloadElInfos()).toMatchSnapshot();

    expect(Global.__FEDERATION__.__PRELOADED_MAP__.size).toBe(1);
    expect(
      Global.__FEDERATION__.__PRELOADED_MAP__.get('@federation/sub3/add'),
    ).toBe(true);

    await FMInstance.preloadRemote([
      {
        nameOrAlias: '@federation/sub3',
        resourceCategory: 'all',
        exposes: ['add'],
      },
    ]);
    expect(getPreloadElInfos()).toMatchSnapshot();
    reset();
  });
});
