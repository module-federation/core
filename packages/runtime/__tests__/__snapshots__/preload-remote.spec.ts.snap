// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preload-remote inBrowser > 1 preload with default config 1`] = `
{
  "links": [
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/button.sync.css",
      "rel": "preload",
      "type": "style",
    },
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/button.sync.js",
      "rel": "preload",
      "type": "script",
    },
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/sub1-button/button.sync.js",
      "rel": "preload",
      "type": "script",
    },
  ],
  "scripts": [
    {
      "crossorigin": "anonymous",
      "src": "http://localhost:1111/resources/preload/preload-resource/federation-remote-entry.js",
    },
    {
      "crossorigin": "anonymous",
      "src": "http://localhost:1111/resources/preload/preload-resource/sub1-button/federation-remote-entry.js",
    },
  ],
}
`;

exports[`preload-remote inBrowser > 2 preload with all config  1`] = `
{
  "links": [
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/sub2/button.async.css",
      "rel": "preload",
      "type": "style",
    },
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/sub2/button.sync.css",
      "rel": "preload",
      "type": "style",
    },
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/sub2/button.async.js",
      "rel": "preload",
      "type": "script",
    },
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/sub2/button.sync.js",
      "rel": "preload",
      "type": "script",
    },
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/sub2-button/button.async.js",
      "rel": "preload",
      "type": "script",
    },
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/sub2-button/button.sync.js",
      "rel": "preload",
      "type": "script",
    },
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/sub2-add/add.async.js",
      "rel": "preload",
      "type": "script",
    },
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/sub2-add/add.sync.js",
      "rel": "preload",
      "type": "script",
    },
  ],
  "scripts": [
    {
      "crossorigin": "anonymous",
      "src": "http://localhost:1111/resources/preload/preload-resource/sub2/federation-remote-entry.js",
    },
    {
      "crossorigin": "anonymous",
      "src": "http://localhost:1111/resources/preload/preload-resource/sub2-button/federation-remote-entry.js",
    },
  ],
}
`;

exports[`preload-remote inBrowser > 3 preload with expose config  1`] = `
{
  "links": [
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/sub3/add.sync.js",
      "rel": "preload",
      "type": "script",
    },
  ],
  "scripts": [
    {
      "crossorigin": "anonymous",
      "src": "http://localhost:1111/resources/preload/preload-resource/sub3/federation-remote-entry.js",
    },
  ],
}
`;

exports[`preload-remote inBrowser > 3 preload with expose config  2`] = `
{
  "links": [
    {
      "href": "http://localhost:1111/resources/preload/preload-resource/sub3/add.sync.js",
      "rel": "preload",
      "type": "script",
    },
  ],
  "scripts": [
    {
      "crossorigin": "anonymous",
      "src": "http://localhost:1111/resources/preload/preload-resource/sub3/federation-remote-entry.js",
    },
  ],
}
`;
